File: my-reader-mobile/modules/readium/ios/Streamer/StreamerConfig.swift
import Foundation
import ReadiumShared

/// Open-architecture Streamer configuration (REP-005/006).
///
/// Phase 1 stores the configuration and always opens publications with the
/// default Readium parsers. Custom `PublicationParser` registration (MOBI/AZW3),
/// `onCreatePublication` transforms (CSS/JS injection, custom services), and
/// content protections (LCP/DRM) are wired in Phase 2. The shape is fixed now
/// so the JS API (`streamer.configure`, `format.register`) is stable.
final class StreamerConfig {
  static let shared = StreamerConfig()

  /// Registered custom format parsers, keyed by registration id. Phase 2 will
  /// route sniffed assets to the matching native `PublicationParser`.
  private(set) var formatRegistrations: [FormatRegistrationRecord] = []

  /// Native transform ids to apply via `onCreatePublication` (Phase 2).
  private(set) var transforms: [String] = []

  /// Native content protections to pass to `PublicationOpener` (Phase 2: LCP/DRM).
  var contentProtections: [ContentProtection] = []

  private init() {}

  func apply(_ config: PublicationOpenerConfigRecord) {
    self.transforms = config.transforms ?? []
    if let formats = config.formats {
      var byId = Dictionary(uniqueKeysWithValues: formatRegistrations.map { ($0.id, $0) })
      for format in formats {
        byId[format.id] = format
      }
      formatRegistrations = Array(byId.values)
    }
  }

  func registerFormat(_ registration: FormatRegistrationRecord) {
    var byId = Dictionary(uniqueKeysWithValues: formatRegistrations.map { ($0.id, $0) })
    byId[registration.id] = registration
    formatRegistrations = Array(byId.values)
  }
}

File: my-reader-mobile/modules/readium/android/src/main/java/com/myreader/readium/Streamer/StreamerConfig.kt
package com.myreader.readium.Streamer

import com.myreader.readium.Types.FormatRegistrationRecord
import com.myreader.readium.Types.PublicationOpenerConfigRecord
import org.readium.r2.shared.publication.protection.ContentProtection

/**
 * Open-architecture configuration for the Readium Streamer (REP-005/006).
 *
 * Singleton holding custom format registrations, native onCreatePublication
 * transform ids, and content-protection schemes. `ReaderService` reads
 * `contentProtections` when building the `PublicationOpener`; the transform /
 * format entries are reserved for Phase 2 (CSS/JS injection, MOBI/AZW3 parsers,
 * LCP) — Phase 1 only stores them so the registration API is wired end-to-end.
 */
object StreamerConfig {
  val formatRegistrations = mutableMapOf<String, FormatRegistrationRecord>()
  val transforms = mutableSetOf<String>()
  var contentProtections: List<ContentProtection> = emptyList()
    private set

  fun apply(config: PublicationOpenerConfigRecord) {
    config.transforms?.let {
      transforms.clear()
      transforms.addAll(it)
    }
    config.formats?.forEach { registerFormat(it) }
    // contentProtection (REP-006) reserved — Phase 3.
  }

  fun registerFormat(registration: FormatRegistrationRecord) {
    formatRegistrations[registration.id] = registration
  }
}
